tribits/ci_support/CDashAnalyzeReportRandomFailures.py

```diff
@@ -0,0 +1,339 @@
+
+      testHistoryFilters = \
+        "filtercount=3&showfilters=1&filtercombine=and"+\
+        "&field1=testname&compare1=63&value1="+nonpassingTest['testname']+\
+        "&field2=groupname&compare2=63&value2="+groupNameNormUrl+\
+        "&field3=buildname&compare3=63&value3="+correctedBuildName
+
+      testHistoryQueryFilters = dateUrlField+"&"+testHistoryFilters
+
+      testHistoryCacheFile = testHistoryCacheDir+"/"+\
+        CDQAR.getCompressedFileNameIfTooLong(testNameBuildName+".json", ext=".json")
+
+      print("\n  Creating file to write test history:\n   "+testHistoryCacheFile)
+
+      testHistoryQueryUrl = CDQAR.getCDashQueryTestsQueryUrl(
+        cdashSiteUrl, cdashProjectName, None, testHistoryQueryFilters)
+      testHistoryBrowserUrl = CDQAR.getCDashQueryTestsBrowserUrl(
+        cdashSiteUrl, cdashProjectName, None, testHistoryQueryFilters)
+
+      testHistoryLOD = CDQAR.downloadTestsOffCDashQueryTestsAndFlatten(
+        testHistoryQueryUrl, testHistoryCacheFile, alwaysUseCacheFileIfExists=True)
+
+      print("\n  Size of test history: "+str(len(testHistoryLOD)))
+
+      if printUrlMode == 'all':
+        print("\n  CDash test history browser URL for "+nonpassingTest['testname']+" "+\
+          correctedBuildName+":\n\n"+"   "+testHistoryBrowserUrl)
+        print("\n  CDash test history query URL for "+nonpassingTest['testname']+" "+\
+          correctedBuildName+":\n\n"+"   "+testHistoryQueryUrl)
+
+      if len(testHistoryLOD) < 2:
+        print("\n  Size of test history too small for any comparisons, skipping ...\n")
+        continue
+
+      # B.3) Split full testing history to passed and nonpassed lists of dicts
+      passingTestHistoryLOD = [test for test in testHistoryLOD if test.get("status") == "Passed"]
+      nonpassingTestHistoryLOD = [test for test in testHistoryLOD if test.get("status") == "Failed"]
+      nonpassingSha1Pairs = set()
+
+      print("\n  Num of passing tests in test history: "+str(len(passingTestHistoryLOD)))
+      print("\n  Num of nonpassing tests in test history: "+str(len(nonpassingTestHistoryLOD)))
+
+      buildSummaryCacheDir = testQueriesCacheDir+"/build_summary_cache/"+testNameBuildName
+      createDirsFromPath(buildSummaryCacheDir)
+      # NOTE: There is an argument to be made that test histories should get their own directory
+      # instead of build summaries and that build summaries should live inside of there
+
+      # C.1) Get all nonpassing tests' sha1s into a set
+      for test in nonpassingTestHistoryLOD:
+
+        buildId = self.getBuildIdFromTest(test)
+
+        buildSummaryCacheFile = buildSummaryCacheDir+"/"+buildId
+        buildSummaryQueryUrl = CDQAR.getCDashBuildSummaryQueryUrl(cdashSiteUrl, buildId)
+        buildConfigOutput = self.downloadBuildSummaryOffCDash(
+          buildSummaryQueryUrl, buildSummaryCacheFile, verbose=printUrlMode =='all',
+          alwaysUseCacheFileIfExists=True)['configure']['output']
+
```

Comment:
Needing to call `['configure']['output']` at this top-level driver code seems to be too much detail about how the CDash data-structure works?  A future refactoring might consider providing the wrapper function `downloadConfigureOutputOffCDash()` that takes the same arguments and just calls:

```
  return downloadBuildSummaryOffCDash( ... args ...)['configure']['output']
```

And the function `downloadConfigureOutputOffCDash()` should be in the module `CDashQueryAnalyzeReport.py` as well.